From ce33e3532b2d9f687eaca850b8e134b0ccf68239 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Sun, 25 Nov 2018 11:08:06 +0200
Subject: [PATCH] version: add version from the build system

This adds version from the build system.
Also improves the date and time handling to take all sources into
the account and generate date and time of the build instead of the
compile date time of a single file preboot.c.

Change-Id: I62c60d105f8323ac7447f08b73312721dbfd77a1
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/Makefile b/Makefile
index b2751c1..23a6920 100644
--- a/Makefile
+++ b/Makefile
@@ -28,6 +28,7 @@
 
 PREBOOT_OBJECTS := $(PREBOOT_SRCS_C:.c=.o)
 PREBOOT_OBJECTS += $(PREBOOT_SRCS_S:.S=.o)
+AUTO_GENERATED_FILE := auto_generated.h
 
 all: preboot.bin
 
@@ -37,13 +38,18 @@
 preboot: $(PREBOOT_OBJECTS)
 	$(CC) $(LDFLAGS) -o $@ -Wl,--start-group $^ $(LDLIBS) -Wl,--end-group
 
-%.o: %.c
+$(AUTO_GENERATED_FILE): localversion-habanalabs $(PREBOOT_SRCS_C) $(PREBOOT_SRCS_S)
+	@( printf '#define PREBOOT_VER "%s"\n' '$(shell cut -d- -f2- localversion-habanalabs)' ) > $@
+	@date +'#define PREBOOT_DATE "%d %b %C%y"' >> $@
+	@date +'#define PREBOOT_TIME "%T"' >> $@
+
+%.o: %.c $(AUTO_GENERATED_FILE)
 	$(CC) $(CFLAGS) -o $@ $<
 
-%.o: %.S
+%.o: %.S $(AUTO_GENERATED_FILE)
 	$(CC) $(CFLAGS) -o $@ $<
 
 clean:
-	rm -f $(PREBOOT_OBJECTS)
+	rm -f $(PREBOOT_OBJECTS) $(AUTO_GENERATED_FILE)
 
 .PHONY: clean
\ No newline at end of file
diff --git a/preboot.c b/preboot.c
index 1c3f614..bdc8e88 100644
--- a/preboot.c
+++ b/preboot.c
@@ -10,6 +10,7 @@
 #include <linux/io.h>
 #include <linux/sizes.h>
 #include "mach/hl_regs.h"
+#include "auto_generated.h"
 #include <stddef.h>
 #include <stdbool.h>
 
@@ -459,10 +460,12 @@
 {
 #ifndef TARGET_IS_TESTER
 	hl_preboot_puts("\n");
-	hl_preboot_puts_raw("Preboot (");
-	hl_preboot_puts_raw(__DATE__);
+	hl_preboot_puts_raw("Preboot ");
+	hl_preboot_puts_raw(PREBOOT_VER);
+	hl_preboot_puts_raw(" (");
+	hl_preboot_puts_raw(PREBOOT_DATE);
 	hl_preboot_puts_raw(" - ");
-	hl_preboot_puts_raw(__TIME__);
+	hl_preboot_puts_raw(PREBOOT_TIME);
 	hl_preboot_puts(")\n");
 #endif
 }
